expression/vectorized.go

```diff
@@ -0,0 +1,215 @@
+			return err
+		}
+		if isNull { // all slots are set to null by PreAlloc()
+			return nil
+		}
+		f64s := result.Float64s()
+		if sel == nil {
+			for i := range f64s {
+				f64s[i] = v
+			}
+			result.SetNulls(0, n, false)
+		} else {
+			for _, i := range sel {
+				f64s[i] = v
+				result.SetNull(i, false)
+			}
+		}
+	case types.ETDecimal:
+		result.PreAllocDecimal(n)
+		v, isNull, err := expr.EvalDecimal(ctx, chunk.Row{})
+		if err != nil {
+			return err
+		}
+		if isNull { // all slots are set to null by PreAlloc()
+			return nil
+		}
+		ds := result.Decimals()
+		if sel == nil {
+			for i := range ds {
+				ds[i] = *v
+			}
+			result.SetNulls(0, n, false)
+		} else {
+			for _, i := range sel {
+				ds[i] = *v
+				result.SetNull(i, false)
+			}
+		}
+	case types.ETDatetime, types.ETTimestamp:
+		result.Reset()
+		v, isNull, err := expr.EvalTime(ctx, chunk.Row{})
+		if err != nil {
+			return err
+		}
+		if isNull {
+			for i := 0; i < n; i++ {
+				result.AppendNull()
+			}
+		} else {
+			if sel == nil {
+				for i := 0; i < n; i++ {
+					result.AppendTime(v)
+				}
+			} else {
+				pos := 0
+				for _, i := range sel {
+					for pos < i {
+						result.AppendNull()
+						pos++
+					}
+					result.AppendTime(v)
+					pos++
+				}
+			}
+		}
+	case types.ETDuration:
+		result.Reset()
+		v, isNull, err := expr.EvalDuration(ctx, chunk.Row{})
+		if err != nil {
+			return err
+		}
+		if isNull {
+			for i := 0; i < n; i++ {
+				result.AppendNull()
+			}
+		} else {
+			if sel == nil {
+				for i := 0; i < n; i++ {
+					result.AppendDuration(v)
+				}
+			} else {
+				pos := 0
+				for _, i := range sel {
+					for pos < i {
+						result.AppendNull()
+						pos++
+					}
+					result.AppendDuration(v)
+					pos++
+				}
+			}
+		}
+	case types.ETJson:
+		result.Reset()
+		v, isNull, err := expr.EvalJSON(ctx, chunk.Row{})
+		if err != nil {
+			return err
+		}
+		if isNull {
+			for i := 0; i < n; i++ {
+				result.AppendNull()
+			}
+		} else {
+			if sel == nil {
+				for i := 0; i < n; i++ {
+					result.AppendJSON(v)
+				}
+			} else {
+				pos := 0
+				for _, i := range sel {
+					for pos < i {
+						result.AppendNull()
+						pos++
+					}
+					result.AppendJSON(v)
+					pos++
+				}
+			}
+		}
+	case types.ETString:
+		result.Reset()
+		v, isNull, err := expr.EvalString(ctx, chunk.Row{})
+		if err != nil {
+			return err
+		}
+		if isNull {
+			for i := 0; i < n; i++ {
+				result.AppendNull()
+			}
+		} else {
+			if sel == nil {
+				for i := 0; i < n; i++ {
+					result.AppendString(v)
+				}
+			} else {
+				pos := 0
+				for _, i := range sel {
+					for pos < i {
+						result.AppendNull()
+						pos++
+					}
+					result.AppendString(v)
+					pos++
+				}
+			}
+		}
+	default:
+		return errors.Errorf("unsupported Constant type for vectorized evaluation")
+	}
+	return nil
+}
+
+// resultLength calculates the appropriate length of result for this chunk.
+func resultLength(chk *chunk.Chunk) int {
```

Comment:
how about:
1. s/resultLength/NumEffectiveRows/
2. bind this function to `Chunk` and put the implementation to `NumRows()`?

Reply:
What does `put the implementation to NumRows()` mean?

Reply:
Oh, it's a mistake. I mean we can put the implementation of these two functions together. So that the API users can easily notice the difference between these two functions.

Reply:
Fixed.